# ACC_Project
Intratumor microbiota in adrenocortical carcinoma

# File description
00.Data: input tables for downstream analysis.

01.Comparison-ACC-PCPG: The folder contains the script used to (1) retrieve the microbial abundance, (2) inspect the microbial profiles at different levels of filtering, and (3) compare the intratumoral microbiome difference between ACC and PCPG.

02.Comparison_WGS_RNA-Seq: The folder contains the script used to compare the microbial profile identified from WGS and RNA-Seq data.

03.Microbial_diversity: The folder contains the script used to explore the microbial data, attempting to find associations between the intratumoral microbiome, microbial diversity, and overall survival.

04.Unsupervised_clustering: The folder contains the script used to group ACC patients with unsupervised clustering algorism based on the intratumoral microbiome.

05.Diversity_between_two_clusters: The folder contains the script used to (1) interrogate the alpha and beta diversity between two clusters identified by the unsupervised clustering method, and (2) investigate the potential confounders that may influence the microbial difference between two clusters.

06.Microbial_features: The folder contains the script used to identify microbial signatures.

07.Clinical_Analysis: This folder contains files used on the correlation analysis between microbial subtype and clinical parameters, and phenotypes. Besides, files used on multivariate cox, ROC, and random forest analysis were included.

08.Genomic_Analysis: This folder contains files used on landscape of genomic events.

09.Transcriptomic_Analysis: This folder contains files used on ESTIMATE score, TIL score, DEG analysis, and GSEA analysis.

10.Host_Microbiome_Analysis: This folder contains files used on procrustes analysis and sparseCCA analysis.

metadata: The matedata of 79 ACC patients.

Tables: Intermediate files generated during the bioinformatic analysis

R_scripts: The R script file of all relevant codes in this study. Ordered as each figure appears in the manuscript.
